File: code/go-utils/hostglob/hostglob.go
// There are four operations on host name patterns and sets of host names.
//
// - We can *match* a pattern or multi-pattern against a set of concrete host names, yielding a
//   selection of those host names
// - We can *expand* a pattern or multi-pattern into a set of concrete host names
// - We can *compress* a set of concrete host names into a pattern or multi-pattern
// - We can *split* a multi-pattern into a set of patterns
//
// The following grammar pertains to all of these:
//
//   multi-pattern   ::= pattern ("," pattern)*
//   pattern         ::= pattern-element ("." pattern-element)*
//   pattern-element ::= fragment+
//   fragment        ::= literal | range | wildcard
//   literal         ::= <longest nonempty string of characters not containing "[" or "," or "*" or ".">
//   range           ::= "[" range-elt ("," range-elt)* "]"
//   range-elt       ::= number | number "-" number
//   number          ::= <nonempty string of 0..9, to be interpreted as decimal>
//   wildcard        ::= "*"
//
//   hostname        ::= host-element ("." host-element)*
//   host-element    ::= literal
//
// The following restrictions apply:
//
// - In a range A-B, A must be no greater than B or the pattern is invalid
// - It is not possible to expand a pattern or multi-pattern that contains a wildcard
// - The expansion of the result of compression of a set of hostnames H must yield exactly
//   the set H
// - Compression does not have a unique result and is not required to be optimal
// - However, compressing the list [y,x] and the list [x,y] must yield the same result (modulo the
//   ordering of the names in the result set), this is important for some consumers.
// - The semantics of matching currently follow the semantics of the regular expression expansion of
//   the pattern-element.  Numbers a, b, ..., z in a range mean /(:?a|b|...|z)/.  A wildcard means
//   /[^.]*/.  The expansion of a pattern-element always starts with /^/ and ends with /$/.  Hence
//   "a[1-3]*b" becomes /^a(?:1|2|3)[^.]*b$/.  This can be confusing: "a[1-3]*b" will actually match
//   the host-element "a14b" because the "1" is matched by the disjunction and the "4" is matched by
//   the wildcard.
//
// Note: There are implementations of the algorithms here both in the Rust code (sonarlog) and in
// the JS code (dashboard/hostglob.js).
//
// Note: matching can be exact (the number of pattern-elements must equal the number of
// host-elements in the hostname) or by prefix (there are fewer pattern-elements than
// host-elements).  In both cases the pattern-elements must match the corresponding host-elements,
// from left towards the right.  In particular, if the pattern is "a*" and the host name is "a.b",
// these must not match if the matching is exact.
//
// Note: an argument can be made that glob semantics are better, so that "a[1-3]*b" would not match
// "a14b": The 14 would be considered "a number" and read as such and matched against {1,2,3}, and
// that would fail.

package hostglob

import (
	"errors"
	"fmt"
	"io"
	"regexp"
	"sort"
	"strconv"
	"strings"

	"go-utils/slices"
)

// This takes a <multi-pattern> according to the grammar above and returns a list of individual
// <pattern>s in that list.  It requires a bit of logic because each pattern may contain a fragment
// that contains a comma.

func SplitMultiPattern(s string) ([]string, error) {
	strings := make([]string, 0)
	if s == "" {
		return strings, nil
	}
	insideBrackets := false
	start := -1
	for ix, c := range s {
		if c == '[' {
			if insideBrackets {
				return nil, fmt.Errorf("Illegal pattern: nested brackets")
			}
			insideBrackets = true
		} else if c == ']' {
			if !insideBrackets {
				return nil, fmt.Errorf("Illegal pattern: unmatched end bracket")
			}
			insideBrackets = false
		} else if c == ',' && !insideBrackets {
			if start == -1 {
				return nil, fmt.Errorf("Illegal pattern: Empty host name")
			}
			strings = append(strings, s[start:ix])
			start = -1
		} else if start == -1 {
			start = ix
		}
	}
	if insideBrackets {
		return nil, fmt.Errorf("Illegal pattern: Missing end bracket")
	}
	if start == len(s) || start == -1 {
		return nil, fmt.Errorf("Illegal pattern: Empty host name")
	}
	strings = append(strings, s[start:])
	return strings, nil
}

// This takes a single <pattern> from the grammar above and expands it.  Restriction: The pattern
// must contain no "*" wildcards.

func ExpandPattern(s string) ([]string, error) {
	before, after, has_tail := strings.Cut(s, ".")
	head_expansions, err := expandPatternElement(before)
	if err != nil {
		return nil, err
	}
	if !has_tail {
		return head_expansions, nil
	}

	tail_expansions, err := ExpandPattern(after)
	if err != nil {
		return nil, err
	}
	expansions := []string{}
	for _, h := range head_expansions {
		for _, t := range tail_expansions {
			expansions = append(expansions, h+"."+t)
		}
	}
	return expansions, nil
}

// MT: Constant after initialization; immutable
var noMoreFragments = errors.New("No more fragments")

type wildcard struct{}

func expandPatternElement(s string) ([]string, error) {
	r := strings.NewReader(s)
	fragments := make([]any, 0)
	for {
		fragment, err := parseFragment(r)
		if err != nil {
			if err == noMoreFragments {
				break
			}
			return nil, err
		}
		if _, ok := fragment.(*wildcard); ok {
			return nil, errors.New("Wildcard not allowed in expansions")
		}
		fragments = append(fragments, fragment)
		if len(fragments) > 100 {
			return nil, errors.New("Unlikely hostname pattern")
		}
	}
	if len(fragments) == 0 {
		return nil, errors.New("Empty element")
	}
	tails := []string{""}
	for i := len(fragments) - 1; i >= 0; i-- {
		switch f := fragments[i].(type) {
		case string:
			xs := make([]string, 0, len(tails))
			for _, t := range tails {
				xs = append(xs, f+t)
			}
			tails = xs
		case []uint64:
			xs := make([]string, 0, len(tails)*len(f))
			for _, t := range tails {
				for _, n := range f {
					xs = append(xs, fmt.Sprintf("%d%s", n, t))
				}
			}
			tails = xs
		default:
			panic("???")
		}
	}
	return tails, nil
}

func parseFragment(r *strings.Reader) (any, error) {
	switch c := getc(r); c {
	case 0:
		return nil, noMoreFragments
	case '*':
		return &wildcard{}, nil
	case '[':
		needOne := true
		nodes := []uint64{}
		for {
			if len(nodes) > 50000 {
				return nil, errors.New("Too many elements")
			}
			if eatc(r, ']') {
				if needOne {
					return nil, errors.New("Expected number")
				}
				break
			}
			needOne = false
			n, err := readNumber(r)
			if err != nil {
				return nil, err
			}
			if eatc(r, '-') {
				m, err := readNumber(r)
				if err != nil {
					return nil, err
				}
				if n > m {
					return nil, errors.New("Bad range")
				}
				for n <= m {
					if len(nodes) > 50000 {
						return nil, errors.New("Too many elements")
					}
					nodes = append(nodes, n)
					n++
				}
			} else {
				nodes = append(nodes, n)
			}
			if eatc(r, ',') {
				needOne = true
			} else if eatc(r, ']') {
				ungetc(r, ']')
			} else {
				return nil, errors.New("Unexpected character")
			}
		}
		return nodes, nil
	case ',':
		return nil, errors.New("Unexpected ','")
	case '.':
		return nil, errors.New("Unexpected '.'")
	default:
		literal := string(c)
		for {
			c := getc(r)
			if c == 0 || c == '[' || c == ',' || c == '.' || c == '*' {
				ungetc(r, c)
				break
			}
			literal = literal + string(c)
		}
		return literal, nil
	}
}

// Given a list of valid <hostname>s by the grammar above, return an abbreviated list that uses
// <pattern> syntax where possible.  The patterns will contain no "*" wildcards.
//
// In general, if there are several compressible ranges within the host names we must pick one.  For
// example, for the set {a1.b1, a2.b2} we naively have two ranges if we consider the host-elements
// separately, but a[1,2].b[1,2] is not a valid compression as it names too many hosts.  While there
// are sets of host names that would allow multiple ranges to be compressed (for example, the set
// resulting from the expansion of that incorrect pattern), this is not a special case worth looking
// for.
//
// For simplicity, for host names of the form `a.b.c...` we will not try to compress anything in the
// `b.c...` portion, and within the `a` portions we will try to compress only the rightmost digit
// strings.  This will yield good results in general.

// MT: Constant after initialization; immutable (except for configuration methods).
var withDigitsRe = regexp.MustCompile(`^(.*?)(\d+)(\D*)$`)

func CompressHostnames(hosts []string) []string {
	// Suffixes is a map from `b.c...` portion to `a` portion of name.
	suffixes := make(map[string][]string)
	for _, h := range hosts {
		before, after, _ := strings.Cut(h, ".")
		if probe, ok := suffixes[after]; ok {
			suffixes[after] = append(probe, before)
		} else {
			suffixes[after] = []string{before}
		}
	}

	// Complete host names
	result := make([]string, 0)

	// Compress the first host-elements, catenate with suffixes
	for suffix, firstelts := range suffixes {
		same := make(map[string][]int)
		for _, elt := range firstelts {
			ms := withDigitsRe.FindStringSubmatch(elt)
			if ms == nil {
				result = pushHostName(result, elt, suffix)
				continue
			}
			n, err := strconv.ParseInt(ms[2], 10, 64)
			if err != nil {
				result = pushHostName(result, elt, suffix)
				continue
			}
			name := ms[1] + "," + ms[3]
			if bag, found := same[name]; found {
				same[name] = append(bag, int(n))
			} else {
				same[name] = []int{int(n)}
			}
		}
		for k, v := range same {
			a, b, _ := strings.Cut(k, ",")
			result = pushHostName(result, a+compressRange(v)+b, suffix)
		}
	}

	return result
}

func pushHostName(result []string, elt, suffix string) []string {
	if suffix != "" {
		return append(result, elt+"."+suffix)
	}
	return append(result, elt)
}

func compressRange(xs []int) string {
	if len(xs) == 1 {
		return fmt.Sprintf("%d", xs[0])
	}
	sort.Sort(sort.IntSlice(xs))
	s := ""
	for i := 0; i < len(xs); {
		first := xs[i]
		prev := first
		i++
		for i < len(xs) && xs[i] == prev+1 {
			prev = xs[i]
			i++
		}
		if s != "" {
			s += ","
		}
		if first != prev {
			s += fmt.Sprintf("%d-%d", first, prev)
		} else {
			s += fmt.Sprintf("%d", first)
		}
	}
	s = "[" + s + "]"
	return s
}

// A `HostGlobber` is a matcher of patterns against hostnames.
//
// The matcher holds a number of patterns, added with `insert`.  Each `pattern` is a <pattern> in
// the sense of the grammar referenced above.
//
// The `match_hostname` method attempts to match its argument against the patterns in the matcher,
// returning true if any of them match.
//
// The hostGlobber is immutable and thread-safe: the embedded regexes are defined to be thread-safe
// for concurrent use. cf https://pkg.go.dev/regexp#Regexp:
//
//  A Regexp is safe for concurrent use by multiple goroutines, except for configuration methods,
//  such as Regexp.Longest.
//
// However note that the RegExp machinery has some shared global state internally in a sync.Map, and
// so executing a regexp is not necessarily lock-free.  Hence the hostglobber is also not
// necessarily lock-free.  But I'm guessing that map is not heavily contended, TBD.  It could look
// like it is only used for fairly complex regexes; ours will tend to be simple (single host
// pattern that should require no backtracking).

type HostGlobber struct {
	isPrefixMatcher bool
	matchers        []*regexp.Regexp
}

// Create a new, empty filter.  The flag indicates whether the globbers in the filter could match
// only a prefix of the hostname elements or must match all of them.

func NewGlobber(isPrefixMatcher bool, patterns []string) (*HostGlobber, error) {
	matchers := make([]*regexp.Regexp, 0, len(patterns))
	for _, p := range patterns {
		re, _, err := compileGlobber(p, isPrefixMatcher)
		if err != nil {
			return nil, err
		}
		matchers = append(matchers, re)
	}
	return &HostGlobber{
		isPrefixMatcher: isPrefixMatcher,
		matchers:        matchers,
	}, nil
}

// Return true iff the filter has no patterns.

func (hg *HostGlobber) IsEmpty() bool {
	return len(hg.matchers) == 0
}

func (hg *HostGlobber) String() string {
	return fmt.Sprint(
		slices.Map(hg.matchers, func(re *regexp.Regexp) string {
			return re.String()
		}),
	)
}

// Match s against the patterns and return true iff it matches at least one pattern.

func (hg *HostGlobber) Match(host string) bool {
	for _, re := range hg.matchers {
		if re.Match([]byte(host)) {
			return true
		}
	}
	return false
}

func compileGlobber(p string, prefix bool) (*regexp.Regexp, string, error) {
	in := strings.NewReader(p)
	r := "^"
Parser:
	for {
		if len(r) > 50000 {
			return nil, "", errors.New("Expression too large, use more '*'")
		}
		switch c := getc(in); c {
		case 0:
			break Parser
		case '*':
			r += "[^.]*"
		case '[':
			ns := make([]string, 0)
			for {
				n0, err := readNumberLimited(in, 0xFFFFFFFF)
				if err != nil {
					return nil, "", err
				}
				if eatc(in, '-') {
					n1, err := readNumberLimited(in, 0xFFFFFFFF)
					if err != nil {
						return nil, "", err
					}
					if n0 > n1 {
						return nil, "", errors.New("Invalid range")
					}
					for n0 <= n1 {
						ns = append(ns, fmt.Sprint(n0))
						n0++
						if len(ns) > 10000 {
							return nil, "", errors.New("Range too large, use more '*'")
						}
					}
				} else {
					ns = append(ns, fmt.Sprint(n0))
				}
				if eatc(in, ']') {
					break
				}
				if !eatc(in, ',') {
					return nil, "", errors.New("Expected ','")
				}
			}
			r += "(?:"
			r += strings.Join(ns, "|")
			r += ")"
		case '.', '$', '^', '?', '\\', '(', ')', '{', '}', ']':
			// Mostly these special chars are not allowed in hostnames, but it doesn't hurt to
			// try to avoid disasters.
			r += "\\"
			r += string(c)
		default:
			r += string(c)
		}
	}
	if prefix {
		// Matching a prefix: we need to match entire host-elements, so following a prefix there
		// should either be EOS or `.` followed by whatever until EOS.
		r += "(?:\\..*)?$"
	} else {
		r += "$"
	}
	re, err := regexp.Compile(r)
	if err != nil {
		return nil, "", err
	}
	return re, r, nil
}

// Common operations on a RuneScanner

func readNumberLimited(r io.RuneScanner, max uint64) (uint64, error) {
	n, err := readNumber(r)
	if err != nil {
		return n, err
	}
	if n > max {
		return 0, errors.New("Number out of range")
	}
	return n, nil
}

func readNumber(r io.RuneScanner) (uint64, error) {
	cs := ""
	for {
		c := getc(r)
		if c < '0' || c > '9' {
			ungetc(r, c)
			break
		}
		cs = cs + string(c)
	}
	if cs == "" {
		return 0, errors.New("Expected number")
	}
	return strconv.ParseUint(cs, 10, 64)
}

func eatc(r io.RuneScanner, x rune) bool {
	c := getc(r)
	if c == x {
		return true
	}
	ungetc(r, c)
	return false
}

func getc(r io.RuneScanner) rune {
	c, _, err := r.ReadRune()
	if err == io.EOF {
		return 0
	}
	return c
}

func ungetc(r io.RuneScanner, c rune) {
	if c != 0 {
		r.UnreadRune()
	}
}

File: code/sonarlog/src/hostglob.rs
/// Matcher, expander and compressor for host names.
///
/// This closely follows the grammar and semantics defined in the corresponding Go and JS code, see
/// go-utils/hostglob/hostglob.go and dashboard/hostglob.js.  They must all be kept in sync.

use crate::pattern;

use anyhow::{bail, Result};
use ustr::Ustr;
use regex::Regex;

/// A `HostGlobber` is a matcher of patterns against hostnames.
///
/// The matcher holds a number of patterns, added with `insert`.  Each `pattern` is a <pattern> in
/// the sense of the grammar referenced above.
///
/// The `match_hostname` method attempts to match its argument against the patterns in the matcher,
/// returning true if any of them match.

pub struct HostGlobber {
    // If true, then the patterns are all constructed to match a prefix of a <hostname>.
    is_prefix_matcher: bool,

    // Matcher + source pattern, for posterity.
    matchers: Vec<(regex::Regex, String)>,
}

impl HostGlobber {
    /// Create a new, empty filter.

    pub fn new(is_prefix_matcher: bool) -> HostGlobber {
        HostGlobber {
            is_prefix_matcher,
            matchers: vec![],
        }
    }

    /// Add the pattern to the set of patterns in the matcher.

    pub fn insert(&mut self, pattern: &str) -> Result<()> {
        self.matchers.push(compile_globber(pattern, self.is_prefix_matcher)?);
        Ok(())
    }

    /// Return true iff the filter has no patterns.

    pub fn is_empty(&self) -> bool {
        self.matchers.len() == 0
    }

    /// Match s against the patterns and return true iff it matches at least one pattern.

    pub fn match_hostname(&self, s: &str) -> bool {
        for m in &self.matchers {
            if m.0.is_match(s) {
                return true
            }
        }
        return false
    }
}

fn compile_globber(p: &str, prefix: bool) -> Result<(Regex, String)> {
    let cs = p.chars().collect::<Vec<char>>();
    let mut i = 0usize;
    let mut r = "^".to_string();
    while i < cs.len() {
        if r.len() > 50000 {
            bail!("Expression too large, use more '*'")
        }
        match cs[i] {
            '*' => {
                i += 1;
                r += "[^.]*";
            }
            '[' => {
                i += 1;
                let mut set = vec![];
                loop {
                    let mut n0;
                    (n0, i) = read_int(&cs, i)?;
                    if i < cs.len() && cs[i] == '-' {
                        i += 1;
                        let n1;
                        (n1, i) = read_int(&cs, i)?;
                        if n0 > n1 {
                            bail!("Invalid range");
                        }
                        while n0 <= n1 {
                            set.push(n0.to_string());
                            n0 += 1;
                            if set.len() > 10000 {
                                bail!("Range too large, use more '*'");
                            }
                        }
                    } else {
                        set.push(n0.to_string());
                    }
                    if i < cs.len() && cs[i] == ']' {
                        i += 1;
                        break;
                    }
                    if i >= cs.len() || cs[i] != ',' {
                        bail!("Expected ','");
                    }
                    i += 1;
                }
                r += "(?:";
                r += &set.join("|");
                r += ")";
            }
            '.' | '$' | '^' | '?' | '\\' | '(' | ')' | '{' | '}' | ']' => {
                // Mostly these special chars are not allowed in hostnames, but it doesn't hurt to
                // try to avoid disasters.
                r += "\\";
                r.push(cs[i]);
                i += 1;
            }
            _ => {
                r.push(cs[i]);
                i += 1;
            }
        }
    }
    if prefix {
        // Matching a prefix: we need to match entire host-elements, so following a prefix there
        // should either be EOS or `.` followed by whatever until EOS.
        r += "(?:\\..*)?$"
    } else {
        r += "$";
    }
    Ok((regex::Regex::new(&r)?, r))
}

fn read_int(cs: &[char], mut i: usize) -> Result<(usize, usize)> {
    let first = i;
    let mut n = 0u64;
    while i < cs.len() && cs[i].is_digit(10) {
        n = n*10 + (u32::from(cs[i]) - 48) as u64;
        if n > 0xFFFFFFFF {
            bail!("Number out of range in glob set");
        }
        i += 1;
    }
    if i == first {
        bail!("Invalid number in glob set");
    }
    Ok((n as usize, i))
}

#[test]
fn test_hostfilter1() {
    let mut hf = HostGlobber::new(true);
    hf.insert("ml8").unwrap();
    hf.insert("ml3.hpc").unwrap();

    // Single-element prefix match against this
    assert!(hf.match_hostname("ml8.hpc.uio.no"));

    // Multi-element prefix match against this
    assert!(hf.match_hostname("ml3.hpc.uio.no"));

    let mut hf = HostGlobber::new(false);
    hf.insert("ml4.hpc.uio.no").unwrap();

    // Exhaustive match against this
    assert!(hf.match_hostname("ml4.hpc.uio.no"));
    assert!(!hf.match_hostname("ml4.hpc.uio.no.yes"));
}

#[test]
fn test_hostfilter2() {
    let mut hf = HostGlobber::new(true);
    hf.insert("ml[1-3]*").unwrap();
    assert!(hf.match_hostname("ml1"));
    assert!(hf.match_hostname("ml1x"));
    assert!(hf.match_hostname("ml1.uio"));
}

#[test]
fn test_hostfilter3() {
    let mut hf = HostGlobber::new(false);
    hf.insert("c[1-3]-[2,4]").unwrap();
    assert!(hf.match_hostname("c1-2"));
    assert!(hf.match_hostname("c2-2"));
    assert!(!hf.match_hostname("c2-3"));
}

// TODO: Less trivial test cases

/// `expand_pattern()` takes a single <pattern> from the grammar and expands it into a set of host
/// names without wildcards or ranges.
///
/// Restriction: The pattern must contain no "*" wildcards.

pub fn expand_pattern(p: &str) -> Result<Vec<String>> {
    let elements = p
        .split('.')
        .map(|x| x.to_string())
        .collect::<Vec<String>>();
    let expansions = expand_elements(&elements)?;
    let mut result = vec![];
    for mut exps in expansions {
        if exps.iter().any(|(prefix, _)| *prefix) {
            bail!("Suffix wildcard not allowed in expandable hostname in config file")
        }
        result.push(
            exps.drain(0..)
                .map(|(_, elt)| elt)
                .collect::<Vec<String>>()
                .join("."),
        )
    }
    Ok(result)
}

// TODO: Test cases

fn expand_elements(xs: &[String]) -> Result<Vec<Vec<(bool, String)>>> {
    if xs.len() == 0 {
        Ok(vec![vec![]])
    } else {
        let rest = expand_elements(&xs[1..])?;
        let expanded = pattern::expand_element(&xs[0])?;
        let mut result = vec![];
        for e in expanded {
            for r in &rest {
                let is_prefix = e.ends_with('*');
                let text = if is_prefix {
                    e[..e.len() - 1].to_string()
                } else {
                    e.to_string()
                };
                let mut m = vec![(is_prefix, text)];
                m.extend_from_slice(&r);
                result.push(m);
            }
        }
        Ok(result)
    }
}

#[test]
fn test_expansion() {
    assert!(
        expand_elements(&vec!["hi[1-2]*".to_string(), "ho[3-4]".to_string()])
            .unwrap()
            .eq(&vec![
                vec![(true, "hi1".to_string()), (false, "ho3".to_string())],
                vec![(true, "hi1".to_string()), (false, "ho4".to_string())],
                vec![(true, "hi2".to_string()), (false, "ho3".to_string())],
                vec![(true, "hi2".to_string()), (false, "ho4".to_string())]
            ])
    )
}

/// `compress_hostnames()` takes a list of <hostname> and returns a vector of <pattern> s.t. the
/// expansion of those patterns results in the original input list.

pub fn compress_hostnames(hosts: &[Ustr]) -> Vec<String> {
    // Given names a1.b.c.d and a2.b.c.d this will return a[1,2].b.c.d, ie, it compresses numbers at
    // the end of the first <host-element> for equal prefixes of those elements and equal tails of
    // elements.  This fits the typical host naming on a supercomputer, which is <name>-<number> or
    // <name><number>.  Note the <name> may also contain digits in the former case.
    //
    // TODO: This code is from an earlier iteration of the problem.  It probably wants to use the
    // same logic as the Go code, which can also compress "a1b" and "a2b" into "a[1,2]b".  This code
    // is still correct, though, and effective in most cases.

    // Split into groups of names a.b.c.d whose tails .b.c.d are the same, we will attempt to merge
    // their `a` elements.
    let mut splits = hosts
        .iter()
        .map(|s| s.as_str().split(".").collect::<Vec<&str>>())
        .collect::<Vec<Vec<&str>>>();

    // Sort lexicographically by tail first and then hostname second - this will allow us to group
    // by tail in a single pass and later group by prefix in another pass.
    splits.sort_by(|a, b| {
        let mut i = 1;
        while i < a.len() || i < b.len() {
            if i < a.len() && i < b.len() {
                if a[i] != b[i] {
                    return a[i].cmp(&b[i]);
                }
            } else if i < a.len() {
                return std::cmp::Ordering::Greater;
            } else {
                return std::cmp::Ordering::Less;
            }
            i += 1;
        }
        return a[0].cmp(&b[0]);
    });

    let mut groups: Vec<&[Vec<&str>]> = vec![];
    let mut i = 0;
    while i < splits.len() {
        let mut j = i + 1;
        while j < splits.len() && same_strs(&splits[i][1..], &splits[j][1..]) {
            j = j + 1
        }
        groups.push(&splits[i..j]);
        i = j
    }

    let mut results: Vec<String> = vec![];
    for g in groups.drain(0..) {
        // Each g has a common tail `.b.c.d`.  Group all the `a` elements from group `g` that can be
        // combined.  The `a` elements that can be combined have a common prefix and a set of
        // combinable suffixes, and since we started with a sorted list, the combinable elements of
        // g are consecutive.

        let lim = g.len();
        let mut i = 0;
        while i < lim {
            let mut j = i + 1;
            let mut suffixes = vec![];
            let mut prefix = None;
            while j < lim {
                if let Some(ix) = combinable(g[i][0], g[j][0]) {
                    if suffixes.is_empty() {
                        prefix = Some(g[i][0][..ix].to_string());
                        suffixes.push(g[i][0][ix..].parse::<usize>().unwrap());
                    }
                    suffixes.push(g[j][0][ix..].parse::<usize>().unwrap());
                    j += 1;
                } else {
                    break;
                }
            }
            if suffixes.is_empty() {
                results.push(g[i].join("."));
            } else {
                // combine several
                let s = prefix.unwrap() + &combine(suffixes);
                let mut elts = vec![s.as_str()];
                elts.extend_from_slice(&g[i][1..]);
                results.push(elts.join("."));
            }
            i = j;
        }
    }

    results
}

fn same_strs(a: &[&str], b: &[&str]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    for i in 0..a.len() {
        if a[i] != b[i] {
            return false;
        }
    }
    return true;
}

// It is known that the prefixes can be combined.
fn combine(mut suffixes: Vec<usize>) -> String {
    suffixes.sort();
    let mut s = "[".to_string();
    let mut k = 0;
    while k < suffixes.len() {
        let mut m = k + 1;
        while m < suffixes.len() && suffixes[m] == suffixes[k] + (m - k) {
            m += 1;
        }
        if k > 0 {
            s += ",";
        }
        if m == k + 1 {
            s += &suffixes[k].to_string();
        } else {
            s += &format!("{}-{}", suffixes[k], suffixes[m - 1]);
        }
        k = m;
    }
    s + "]"
}

#[test]
fn test_compress_hostnames() {
    assert!(
        compress_hostnames(&vec![
            Ustr::from("a1"),
            Ustr::from("a3"),
            Ustr::from("a2"),
            Ustr::from("a5")
        ]).join(",") == "a[1-3,5]"
    );
    // Hosts are carefully ordered here to ensure that they are not sorted either by their first or
    // second elements.
    assert!(
        compress_hostnames(&vec![
            Ustr::from("a3.fox"),
            Ustr::from("a1.fox"),
            Ustr::from("a3.fum"),
            Ustr::from("a2.fox"),
            Ustr::from("a5.fox"),
        ]).join(",") == "a[1-3,5].fox,a3.fum"
    );
}

// Names can be merged if they both end with a digit string and there is a joint prefix before the
// digit string.  For now, we require this prefix to not end with a digit.  This returns None for
// "no" and Some(usize) for "yes" where usize is the byte index of the start of the digit string.
//
// Combinability must be reflexive, symmetric, and transitive.

fn combinable(a: &str, b: &str) -> Option<usize> {
    let xs = a.as_bytes();
    let mut i = xs.len();
    while i > 0 && xs[i - 1] >= b'0' && xs[i - 1] <= b'9' {
        i -= 1;
    }
    if i == 0 || i == xs.len() {
        return None;
    }
    let ys = b.as_bytes();
    let mut j = ys.len();
    while j > 0 && ys[j - 1] >= b'0' && ys[j - 1] <= b'9' {
        j -= 1;
    }
    if j == 0 || j == ys.len() {
        return None;
    }
    if i != j {
        return None;
    }
    if xs[..i] != ys[..j] {
        return None;
    }
    return Some(i);
}

#[test]
fn test_elements_can_be_merged() {
    assert!(combinable("", "") == None);
    assert!(combinable("a", "b") == None);
    assert!(combinable("a", "a") == None);
    assert!(combinable("a1", "a23") == Some(1));
    assert!(combinable("a1-1", "a1-23") == Some(3));
}
